rclcpp/include/rclcpp/experimental/serialization.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__EXPERIMENTAL__SERIALIZATION_HPP_
+#define RCLCPP__EXPERIMENTAL__SERIALIZATION_HPP_
+
+#include <rmw/rmw.h>
+
+#include <memory>
+#include <string>
+
+#include "rcl/error_handling.h"
+
+namespace rclcpp
+{
+namespace experimental
+{
+
+/// Interface to (de)serialize a message
+class SerializationBase
+{
+public:
+  virtual ~SerializationBase() = default;
+
+  /// Serialize a ROS2 message to a serialized stream
+  /**
+   * \param[in] message The ROS2 message which is read and serialized by rmw.
+   */
+  virtual std::shared_ptr<rcl_serialized_message_t> serialize_message(const void * message) = 0;
+
+  /// Deserialize a serialized stream to a ROS message
+  /**
+   * \param[in] serialized_message The serialized message to be converted to ROS2 by rmw.
+   * \param[out] message The deserialized ROS2 message.
+   */
+  virtual void deserialize_message(
+    const rcl_serialized_message_t * serialized_message,
+    void * msg) = 0;
+};
+
+/// Default implementation to (de)serialize a message by using rmw_(de)serialize
+class Serialization : public SerializationBase
+{
+public:
+  Serialization(
+    const rosidl_message_type_support_t & type_support,
+    const rcutils_allocator_t allocator = rcutils_get_default_allocator())
+  : type_support_(type_support), rcutils_allocator_(allocator)
+  {}
+
+  std::shared_ptr<rcl_serialized_message_t> serialize_message(const void * message) override
+  {
+    auto serialized_message = new rcl_serialized_message_t;
+    *serialized_message = rmw_get_zero_initialized_serialized_message();
+    const auto ret = rmw_serialized_message_init(serialized_message, 0, &rcutils_allocator_);
+    if (ret != RCUTILS_RET_OK) {
+      throw std::runtime_error(
+              "Error allocating resources for serialized message: " +
+              std::string(rcutils_get_error_string().str));
+    }
+
+    if (nullptr == message) {
+      throw std::runtime_error("Message is nullpointer while serialization.");
+    }
```

Comment:
still leaking memory here. This check should be first thing in this function, before any allocation is done